src/main/java/io/github/tsegismont/vertx/contextual/logging/JULContextualDataFormatter.java

```diff
@@ -0,0 +1,186 @@
+import java.util.logging.LogRecord;
+
+/**
+ * JUL Formatted that is able to process Vert.x Context data.
+ * @author Paulo Lopes
+ */
+public final class JULContextualDataFormatter extends Formatter {
+
+  private static final String placeholderPrefix = "%{";
+  private static final String placeholderSuffix = "}";
+
+  private final String template;
+
+  private final Date dat = new Date();
+
+  private static final List<String> RESERVED = Arrays.asList("date", "source", "logger", "level", "message", "thrown");
+
+  private final List<BiFunction<LogRecord, ContextInternal, Object>> resolvers = new ArrayList<>();
+
+  public JULContextualDataFormatter() {
+    this(LogManager.getLogManager().getProperty("format"));
+  }
+
+  public JULContextualDataFormatter(String template) {
+    // add the default resolvers
+    // 1. date
+    resolvers.add((record, ctx) -> {
+      // with java 11 this should be replaced with the new time APIs
+      dat.setTime(record.getMillis());
+      return dat;
+    });
+    // 2. source
+    resolvers.add((record, ctx) -> {
+      String source;
+      if (record.getSourceClassName() != null) {
+        source = record.getSourceClassName();
+        if (record.getSourceMethodName() != null) {
+          source = source + " " + record.getSourceMethodName();
+        }
+      } else {
+        source = record.getLoggerName();
+      }
+      return source;
+    });
+    // 3. logger
+    resolvers.add((record, ctx) -> record.getLoggerName());
+    // 4. level
+    resolvers.add((record, ctx) -> record.getLevel().getLocalizedName());
+    // 5. message
+    resolvers.add((record, ctx) -> formatMessage(record));
+    // 6. thrown
+    resolvers.add((record, ctx) -> {
+      String throwable = "";
+      if (record.getThrown() != null) {
+        StringWriter sw = new StringWriter();
+        PrintWriter pw = new PrintWriter(sw);
+        pw.println();
+        record.getThrown().printStackTrace(pw);
+        pw.close();
+        throwable = sw.toString();
+      }
+      return throwable;
+    });
+
+    this.template = parseStringValue(template);
+  }
+
+  private String parseStringValue(String template) {
+
+    StringBuilder buf = new StringBuilder(template);
+
+    int startIndex = template.indexOf(placeholderPrefix);
+    while (startIndex != -1) {
+      int endIndex = findPlaceholderEndIndex(buf, startIndex);
+      if (endIndex != -1) {
+        String placeholder = buf.substring(startIndex + placeholderPrefix.length(), endIndex);
+
+        int index = RESERVED.indexOf(placeholder);
+
+        if (index == -1) {
+          // lookup default value
+          int defIndex = placeholder.indexOf(":-");
+
+          // need to lock to use inside lambda
+          final String defValue;
+          final String ctxKey;
+
+          if (defIndex != -1) {
+            ctxKey = placeholder.substring(0, defIndex);
+            defValue = placeholder.substring(defIndex + 2);
+          } else {
+            defValue = "-";
```

Comment:
The other converters use an empty string when no default is provided. How about doing the same for consistency?